### PR TITLE
Improve mail sender using private fs over mq

### PR DESCRIPTION
**Pull request by ghost**
### 1. Why is this change necessary?
- Using the message queue with a transport such as SQS is problematic for large email messages because a transport such as SQS only supports messages up to 256 KB in size.
- This PR moves the email data away from the message queue, stores it in the private file system and sends only the path to the private file in a queue message.
- This allows us to bypass the transport size limit and still support low latency because we send the email only with from the consumer / handler of the final message.

### 2. What does this change do, exactly?
* Changed `src/Core/Content/Mail/Service/AbstractMailSender.php` to deprecate the `envelope` parameter
* Changed `src/Core/Content/Mail/Service/MailSender.php` to no longer directly send the mail to the symfony mailer, but to write the serialized mail to the private file system & dispatch a `SendMailMessage` to the message bus
* Added `src/Core/Content/Mail/Message/SendMailMessage.php` to hold the `mailDataPath` on the message bus
* Added `src/Core/Content/Mail/Message/SendMailHandler.php` to consume the `SendMailMessage`, send the mail & delete the serialized mail from the private file system 

### 3. Describe each step to reproduce the issue or behaviour.
- Send a large email object over the message queue (e.g. with attachments)
- Transports like SQS can't handle the message as it is to big

### 4. Please link to the relevant issues (if any).
- #5462

### 5. Checklist

- [X] I have rebased my changes to remove merge conflicts
- [X] I have written tests and verified that they fail without my change
- [X] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [X] I have written or adjusted the documentation according to my changes
- [X] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfill them.

### CODE REVIEW

File: src/Core/Content/Mail/Message/SendMailHandler.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+            return;
+        }
+
+        $this->transport->send($mail);
+
+        try {
+            $this->filesystem->delete($mailDataPath);
+        } catch (\Throwable) {
+            // Ignore any potential errors while deleting the file, as a error would reschedule the message and send the mail again
```
Comment (tinect) [inline src/Core/Content/Mail/Message/SendMailHandler.php]: We should log it, shouldn't we? Or throw a [UnrecoverableMessageHandlingException](https://symfony.com/doc/current/messenger.html#avoiding-retrying)

File: src/Core/Content/Mail/Message/SendMailHandler.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        if (!($mail instanceof Email)) {
+            $this->filesystem->delete($mailDataPath);
+
+            $this->logger->error('An error occurred while sending an email. The mail data is invalid.');
+
+            return;
+        }
+
+        $this->transport->send($mail);
```
Comment (tinect) [inline src/Core/Content/Mail/Message/SendMailHandler.php:55]: The envelope is gone?

**Review by tinect — CHANGES_REQUESTED** [review]

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 1 above)
Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: ~~As the commet says: If we throw a error here the message would be reschedule, what is not what we want after we already send the mail.~~
But yes we can log it.

File: src/Core/Content/Mail/Message/SendMailHandler.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * @throws TransportExceptionInterface
+     */
+    public function __invoke(SendMailMessage $message): void
+    {
+        $mailDataPath = $message->getMailDataPath();
+
+        if (!$this->filesystem->fileExists($mailDataPath)) {
```
Comment (jenskueper) [inline src/Core/Content/Mail/Message/SendMailHandler.php]: Ideally, this call can be omitted and an exception catched/handled during the actual read to reduce the overall cost for distributed filesystems where every file operation is billed.

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 2 above)
Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php:55]: The MailSender has a internal construct and is only used by the MailService which never passes the envelope object. So I guess we don't need it anymore as it was never used before.

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 1 above)
Comment (jenskueper) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: "If you throw UnrecoverableMessageHandlingException, the message will not be retried."

Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: Yep, sorry, I clicked on the link too late. I read it and am now adjusting it.

File: src/Core/Content/Mail/Message/SendMailHandler.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            $this->logger->error('An error occurred while sending an email. The mail data file does not exist.');
```
Comment (keulinho) [inline src/Core/Content/Mail/Message/SendMailHandler.php]: we should include the path here for context to make debugging easier, and probably also document why we don't throw here as it does not make sense to retry that message right? 

Or should we rather use an exception and mark it as non-retryable? https://symfony.com/doc/current/messenger.html#avoiding-retrying

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 1 above)
Comment (keulinho) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: ah commented something similiar on the other log statements

But is there any reason not using the UnrecoverableMessageHandlingException? as that would make debugging a lot easier as the failed messages should be persistent in the failed transport and could be inspected manually 🤔 

at least for everything before message sending, when we send the mail I guess it is correct to let the handler exit successfully even if we can not clean up

Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: I changed it.

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 3 above)
Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: I changed it.

File: src/Core/Content/Mail/Message/SendMailHandler.php (hunk 4 above)
Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: I included the path in the error message.

File: src/Core/Content/Mail/Message/SendMailHandler.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+     * @throws UnrecoverableMessageHandlingException
+     */
+    public function __invoke(SendMailMessage $message): void
+    {
+        $mailDataPath = $message->getMailDataPath();
+
+        try {
+            $mailData = $this->filesystem->read($mailDataPath);
+        } catch (\Throwable $e) {
+            throw new UnrecoverableMessageHandlingException(\sprintf('An error occurred while reading the mail data file "%s".', $mailDataPath), 0, $e);
```
Comment (jenskueper) [inline src/Core/Content/Mail/Message/SendMailHandler.php]: Can you distinguish between file does not exist and other errors. File does not exist would be an unrecoverable, others e.g. caused by temporary network issues would make sense to retry them.

Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: As far as I can see in the FilesystemAdapters of e.g. Local or S3, they all just catch all errors and throw a UnableToReadFile themselves. So we will always get the same error. Unless we explicitly check if the file exists beforehand, we have no way of knowing what the reason for the error is.
The file could be temporarily or permanently gone.

Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: @keulinho @shyim What do you think about this & about https://github.com/shopware/shopware/pull/5490#discussion_r1841976012?

Comment (tinect) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: Yes, they just throw UnableToReadFile exceptions. Why not safety, let them retry on read-issues?

Comment (tinect) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: But I would make the message more clear by adding the information, that the mail has not been sent.

Comment (Gaitholabi) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: What I would do, is place the fileExists call inside the catch exception block. In that case, the 2 calls to the storage would be made only in edge cases where we need to be sure that the error is recoverable or not

Comment (ghost) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: @Gaitholabi I added the check and more tests.

Comment (Gaitholabi) [reply src/Core/Content/Mail/Message/SendMailHandler.php]: hey, a quick update, we're still looking at this and looking to get it done soon. thanks!

**Comment by Gaitholabi** [comment]
@akf-bw Thank you for your contribution! this has been merged

**Comment by ghost** [comment]
@Gaitholabi What is the reasoning behind the decision to remove "UnrecoverableMessageHandlingException"?
Also why do you try to rerun the task handler, when the mail could not be unserialize correctly anymore?
Your changes will put more load on the filesystem in places where we don't have to do this.
And why use the hasher instead of Uuid? Using a hash on such a large string as a serialized mail object takes significantly longer and reduces performance, which is not what we actually wanted with these changes.
You made some changes to my PR that are pretty unfounded for me and have made the actual implementation a bit worse. Could you please explain why you changed my code in those places? Thanks!

**Comment by Gaitholabi** [comment]
> @Gaitholabi What is the reasoning behind the decision to remove "UnrecoverableMessageHandlingException"? Also why do you try to rerun the task handler, when the mail could not be unserialize correctly anymore? Your changes will put more load on the filesystem in places where we don't have to do this. And why use the hasher instead of Uuid? Using a hash on such a large string as a serialized mail object takes significantly longer and reduces performance, which is not what we actually wanted with these changes. You made some changes to my PR that are pretty unfounded for me and have made the actual implementation a bit worse. Could you please explain why you changed my code in those places? Thanks!

Hello @akf-bw,

Apologies for the delay in the reply. I'm happy to clarify the changes.

1. In the case of a failed serialization, the handler successfully finishes, and there should not be any retry attempts. We just log an error.
2. With `UnrecoverableMessageHandlingException`,  those messages would have ended up in the [failed queue](https://symfony.com/doc/current/messenger.html#avoiding-retrying). in the given exceptions, we are absolutely sure that retrying those later will not help. So we just log and error, and let the handler finish successfully.
3. Regarding filesystem changes: in normal operation (should be always in a healthy system), there are only two filesystem calls—one for reading and one for cleanup. Could you point out where you’re seeing extra load? There was no modification to the PR in that area.
4. Switching from UUID to hashing creates predictable filenames, which improves reproducibility in tests and simplifies filesystem management. While hashing is slower, the performance overhead is minimal—over 10,000 runs, hashing 512 KB took about 200 ms, compared to roughly 30 ms for generating a random UUID.

**Comment by ghost** [comment]
@Gaitholabi I must have misread the code. When I saw the `logger->error` ago I thought it was an exception. Don't know why I misread that there.

> 4\. Switching from UUID to hashing creates predictable filenames, which improves reproducibility in tests and simplifies filesystem management. While hashing is slower, the performance overhead is minimal—over 10,000 runs, hashing 512 KB took about 200 ms, compared to roughly 30 ms for generating a random UUID.

You can use them in the tests without problems, also Uuid v7 creates a sortable list in the file system (old -> new files). I don't know where the hash is supposed to "simplifies filesystem management", so I still disagree with this decision.

Also, the cleanup functionality is now missing for the part where the email couldn't be serialized properly, so it would currently create dead files in the filesystem. I'll create a PR to fix this for you.

**Comment by Gaitholabi** [comment]
@akf-bw Leaving the file intact is intentional. That would help finding the error, in case that ever happens. 

For filesystem management, the simplification mainly comes from predictability. For example, in the case of tests, you can 'expect' the filename based on the corresponding file content. In either case, there isn't a decisively compelling argument favoring one approach over the other. A hash can still be calculated in real-time later, and the performance difference is negligible unless most of the files are large, and even then the impact is still marginal.
